File: pachi/pachi-rust/src/hopper.rs
use core::num;

use godot::{classes::Timer, prelude::*};

use crate::{ball::Ball, ball_source::BallSource, game::Game};

/// Pachinko Board
#[derive(GodotClass)]
#[class(init, base=Node2D)]
pub struct Hopper {
    #[export]
    launch_sources: Array<Gd<BallSource>>,

    #[export]
    launch_timer: Option<Gd<Timer>>,

    #[export]
    default_ball_scn: Option<Gd<PackedScene>>,

    #[export]
    #[init(val = 20)]
    initial_balls: u32,

    #[init(val = 0)]
    next_launch_source: usize,

    pending_balls: Array<Gd<Ball>>,

    base: Base<Node2D>,
}

#[godot_api]
impl INode2D for Hopper {
    fn ready(&mut self) {
        assert!(self.launch_timer.is_some());
        assert!(self.default_ball_scn.is_some());

        self.launch_timer
            .as_ref()
            .unwrap()
            .signals()
            .timeout()
            .connect_other(&mut self.to_gd(), |hopper| hopper.try_launch_next_ball());

        self.add_default_balls(self.initial_balls as usize);

        let game = Game::autoload();
        game.bind()
            .events
            .signals()
            .add_default_balls()
            .connect_other(&self.to_gd(), |inner_self, num_balls| {
                inner_self.add_default_balls(num_balls as usize)
            });
    }
}

#[godot_api]
impl Hopper {
    fn try_launch_next_ball(&mut self) {
        let next_ball = match self.pending_balls.pop() {
            Some(ball) => ball,
            _ => return,
        };

        assert!(self.next_launch_source < self.launch_sources.len());
        let mut launch_source: Gd<BallSource> =
            self.launch_sources.get(self.next_launch_source).unwrap();
        launch_source.bind_mut().launch_existing_ball(next_ball, 1.);

        self.next_launch_source = (self.next_launch_source + 1) % self.launch_sources.len();
    }

    pub fn add_default_balls(&mut self, num_balls: usize) {
        for _ in 0..num_balls {
            let ball = self
                .default_ball_scn
                .as_ref()
                .unwrap()
                .instantiate()
                .unwrap()
                .cast::<Ball>();
            self.pending_balls.push(&ball);
        }
    }

    // TODO: bookkeeping to make this less expensive
    #[func]
    pub fn ball_count(&self) -> u32 {
        let mut out = 0;
        for child in self.base().get_children().iter_shared() {
            match child.try_cast::<Ball>() {
                Ok(_) => out += 1,
                _ => {}
            }
        }
        out
    }

    pub fn destroy_balls(&mut self, num_balls: usize) {
        self.base_mut()
            .get_children()
            .iter_shared()
            .filter(|child| match child.clone().try_cast::<Ball>() {
                Ok(_) => true,
                _ => false,
            })
            .take(num_balls)
            .for_each(|mut child| child.queue_free());
    }
}

File: pachi/content/hopper/Hopper.cs
using Godot;
using System.Collections.Generic;
using System.Linq;

public partial class Hopper : Node2D
{
    [Export]
    public Godot.Collections.Array<BallSource> LaunchSources { get; set; } = new Godot.Collections.Array<BallSource>();

    [Export]
    public Timer LaunchTimer { get; set; }

    [Export]
    public PackedScene DefaultBallScn { get; set; }

    [Export]
    public uint InitialBalls { get; set; } = 20;

    [Export]
    public uint MaxBalls { get; set; } = 256;

    private int _nextLaunchSource = 0;
    private Queue<Ball> _pendingBalls = new Queue<Ball>();
    private Queue<Ball> _containedBalls = new Queue<Ball>();

    public override void _Ready()
    {
        if (LaunchTimer == null) GD.PushError("Hopper: LaunchTimer is not assigned!");
        if (DefaultBallScn == null) GD.PushError("Hopper: DefaultBallScn is not assigned!");

        LaunchTimer?.Connect(Timer.SignalName.Timeout, Callable.From(TryLaunchNextBall));

        AddDefaultBalls((int)InitialBalls);

        Game.Instance.Events.AddDefaultBalls += OnAddDefaultBalls;
    }

    private void OnAddDefaultBalls(int numBalls)
    {
        AddDefaultBalls(numBalls);
    }

    private void TryLaunchNextBall()
    {
        if (_pendingBalls.Count == 0) return;
        if (_containedBalls.Count >= MaxBalls) return;

        _nextLaunchSource = (_nextLaunchSource + 1) % LaunchSources.Count;
        // Ball nextBall = _pendingBalls[_pendingBalls.Count - 1];
        Ball nextBall = _pendingBalls.Dequeue();
        BallSource source = LaunchSources[_nextLaunchSource];
        _containedBalls.Enqueue(nextBall);
        source.LaunchExistingBall(nextBall, 1.0f);
    }

    public void AddDefaultBalls(int numBalls)
    {
        if (DefaultBallScn == null) return;

        for (int i = 0; i < numBalls; i++)
        {
            AddBall(DefaultBallScn.Instantiate<Ball>());
        }
    }

    public void AddBall(Ball ball)
    {
        // TODO: if we are at max balls, convert ball to resources
        _pendingBalls.Enqueue(ball);
    }

    public uint GetBallCount()
    {
        uint count = 0;
        foreach (Node child in GetChildren())
        {
            if (child is Ball) count++;
        }
        return count;
    }

    public void DestroyBalls(int numBalls)
    {
        var balls = GetChildren().OfType<Ball>().Take(numBalls).ToList();
        foreach (var ball in balls)
        {
            ball.QueueFree();
        }
    }

    public Ball DequeueNextBall()
    {
        if (_containedBalls.Count == 0) return null;
        return _containedBalls.Dequeue();
    }
}
